### PR TITLE
ARROW-5765: [C++] Fix TestDictionary.Validate in release mode, add docker-compose job for testing C++ release build

### PR DESCRIPTION
**Pull request by wesm**
This was exposed by ARROW-5145 https://github.com/apache/arrow/commit/f77c3427ca801597b572fb197b92b0133269049b

I also added a docker-compose job for testing the release build nightly at least in Crossbow on https://github.com/ursa-labs/crossbow

### CODE REVIEW

**Comment by wesm** [comment]
@ursabot crossbow test docker-cpp-release

**Comment by wesm** [comment]
@ursabot crossbow test docker-cpp-release

**Comment by wesm** [comment]
@ursabot crossbow test docker-cpp-release

**Comment by wesm** [comment]
@kszucs @fsaintjacques it seems like ursabot is crashing or something. any ideas?

**Comment by wesm** [comment]
I verified that the docker-compose job passes locally

**Comment by wesm** [comment]
@pitrou we can use `ASSERT_DEATH` to write unit tests that hit `ARROW_CHECK_OK` or `DCHECK` paths. We should perhaps start being more diligent about writing such tests to check assertions are enforced

**Comment by wesm** [comment]
+1

File: docker-compose.yml [hunk 1]
```diff
@@ -118,6 +118,20 @@ services:
...
+    # Usage:
+    #   docker-compose build cpp
+    #   docker-compose run cpp-release
+    image: arrow:cpp
+    shm_size: 2G
+    build:
+      context: .
+      dockerfile: cpp/Dockerfile
+    environment:
+      ARROW_BUILD_TYPE: release
```
Comment (pitrou) [inline docker-compose.yml]: Does it get passed automatically to CMake?

Comment (wesm) [reply docker-compose.yml]: see https://github.com/apache/arrow/blob/master/ci/docker_build_cpp.sh#L34

**Pull request closed without merging**